File: smart_farm/src/api/farmTask.ts
import { get, http, post } from '@/utils/request'
import type { PageInfo } from '@/api/user'

/** 农事任务实体，对应 smart_plant 后端 FarmTask 返回结构。 */
export interface FarmTask {
  /** 任务ID，对应 farm_task.id。 */
  id: number
  /** 任务归属用户ID，农事消息视角也会作为发布人ID使用。 */
  userId: number
  /** 归属用户账号。 */
  username?: string
  /** 归属用户昵称。 */
  nickname?: string
  /** 农事消息视角使用的发布人名称，优先昵称、兜底账号。 */
  publisherName?: string
  /** 任务所属地块ID。 */
  plotId: number
  /** 地块名称。 */
  plotName?: string
  /** 地块编号，用于辅助区分同名地块。 */
  plotCode?: string
  /** 种植批次ID。 */
  batchId?: number
  /** 种植批次编号。 */
  batchNo?: string
  /** 任务标题，支持条件查询。 */
  taskTitle: string
  /** 任务类型：1浇水 2施肥 3打药 4采收 5巡检 6除草 7补光 8其他。 */
  taskType: number
  /** 任务内容，农事消息页面也会按该字段查询。 */
  taskContent?: string
  /** 优先级：1低 2普通 3高 4紧急。 */
  priority?: number
  /** 任务截至时间。 */
  deadlineTime?: string
  /** 实际开始时间。 */
  actualStartTime?: string
  /** 实际完成时间。 */
  actualEndTime?: string
  /** 任务状态：1未开始 2进行中 3已完成 4已逾期 5已取消。 */
  status?: number
  /** 执行人ID。 */
  executorId?: number
  /** 执行人名称。 */
  executorName?: string
  /** 完成说明。 */
  completeRemark?: string
  /** 后台备注。 */
  remark?: string
  /** 来源类型，例如 AI_SOLUTION。 */
  sourceType?: string
  /** 来源业务ID。 */
  sourceId?: number
  /** AI 处置方案ID。 */
  aiSolutionId?: number
  /** 创建时间。 */
  createTime?: string
  /** 发布时间别名，后端由 create_time 映射。 */
  publishTime?: string
  /** 更新时间。 */
  updateTime?: string
  /** 是否已超过截至时间，该值由后端动态计算。 */
  overdue?: boolean
}

/** 农事任务列表查询参数，同时兼容农事消息页面的标题、内容、地块查询。 */
export interface FarmTaskListParams {
  plotName?: string
  taskTitle?: string
  taskContent?: string
  taskType?: number
  status?: number
  sourceType?: string
  startDate?: string
  endDate?: string
  pageNum: number
  pageSize: number
}

/** 新增或编辑农事任务时提交的表单字段。 */
export interface FarmTaskPayload {
  plotId: number
  batchId?: number
  taskTitle: string
  taskType: number
  taskContent?: string
  priority?: number
  /** 新增或编辑任务时设置的单一截至时间。 */
  deadlineTime?: string
  /** 可预先指派执行人；实际领取任务时后端会记录最终执行人。 */
  executorId?: number
  remark?: string
}

export interface UpdateFarmTaskPayload extends Partial<FarmTaskPayload> {
  id: number
}

/** smart_plant 统一响应包装。 */
interface ApiResult<T> {
  code: number
  message: string
  data: T
}

/** 查询农事任务详情，供详情抽屉和编辑弹窗回填。 */
export const getFarmTaskById = (id: number) => {
  return get<ApiResult<FarmTask>>(`/farm-task/${id}`)
}

/** 新增农事任务，对接 smart_plant POST /farm-task/add。 */
export const addFarmTask = (data: FarmTaskPayload) => {
  return post<ApiResult<FarmTask>>('/farm-task/add', data as unknown as Record<string, unknown>)
}

/** 编辑农事任务，对接 smart_plant PUT /farm-task。 */
export const updateFarmTask = (data: UpdateFarmTaskPayload) => {
  return http<ApiResult<FarmTask>>({ url: '/farm-task', method: 'put', data })
}

/** 删除单条农事任务。 */
export const deleteFarmTask = (id: number) => {
  return http<ApiResult<void>>({ url: `/farm-task/${id}`, method: 'delete' })
}

/** 批量删除农事任务。 */
export const batchDeleteFarmTasks = (ids: number[]) => {
  return http<ApiResult<number>>({ url: '/farm-task/batch', method: 'delete', data: ids })
}

/** 更新农事任务执行状态，详情抽屉中的开始/完成按钮会调用该接口。 */
export const updateFarmTaskStatus = (id: number, status: number, completeRemark?: string) => {
  return http<ApiResult<FarmTask>>({
    url: `/farm-task/${id}/status`,
    method: 'put',
    params: { status, completeRemark },
  })
}

/** 分页查询农事任务，参数名与后端 FarmTaskController 保持一致。 */
export const listFarmTasks = (params: FarmTaskListParams) => {
  return get<ApiResult<PageInfo<FarmTask>>>('/farm-task/list', params as unknown as Record<string, unknown>)
}

/** 查询农事任务统计卡片数据。 */
export const getFarmTaskStatistics = () => {
  return get<ApiResult<Record<string, number>>>('/farm-task/statistics')
}

File: smart_plant/src/main/java/com/smart_plant/smart_plant/entity/FarmTask.java
package com.smart_plant.smart_plant.entity;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class FarmTask {

    /** 农事任务主键ID，对应 farm_task.id。 */
    private Long id;

    /** 农事归属用户ID，后台列表中同时作为发布人ID使用。 */
    private Long userId;

    /** 发布人登录账号，来自 user.username。 */
    private String username;

    /** 发布人昵称，优先用于页面展示。 */
    private String nickname;

    /** 农事所属地块ID，对应 farm_task.plot_id。 */
    private Long plotId;

    /** 地块名称，列表查询时由 plot 表关联得到。 */
    private String plotName;

    /** 地块编号，便于运营人员区分同名地块。 */
    private String plotCode;

    /** 种植批次ID，用于把农事和当前作物批次关联。 */
    private Long batchId;

    /** 批次编号，详情展示时由 planting_batch 表关联得到。 */
    private String batchNo;

    /** 农事标题，是管理端列表和条件查询的核心字段。 */
    private String taskTitle;

    /** 农事类型：1浇水 2施肥 3打药 4采收 5巡检 6除草 7补光 8其他。 */
    private Integer taskType;

    /** 农事内容，支持管理端按内容关键词检索。 */
    private String taskContent;

    /** 优先级：1低 2普通 3高 4紧急。 */
    private Integer priority;

    /** 任务截至时间，替代容易产生歧义的计划开始/计划结束区间。 */
    private LocalDateTime deadlineTime;

    /** 实际开始时间，由状态流转到进行中时自动记录。 */
    private LocalDateTime actualStartTime;

    /** 实际完成时间，由状态流转到已完成时自动记录。 */
    private LocalDateTime actualEndTime;

    /** 农事状态：1未开始 2进行中 3已完成 4已逾期 5已取消。 */
    private Integer status;

    /** 执行人用户ID。 */
    private Long executorId;

    /** 执行人名称，列表查询时由 user 表关联得到。 */
    private String executorName;

    /** 完成备注，用于记录执行闭环说明。 */
    private String completeRemark;

    /** 后台备注，不直接参与业务状态流转。 */
    private String remark;

    /** 来源类型，例如 AI_SOLUTION 表示由 AI 处置方案生成。 */
    private String sourceType;

    /** 来源业务ID，与 sourceType 共同定位来源记录。 */
    private Long sourceId;

    /** AI 处置方案ID，便于从识别结果追溯到农事任务。 */
    private Long aiSolutionId;

    /** 创建时间，管理端作为农事发布时间展示。 */
    private LocalDateTime createTime;

    /** 发布时间别名，便于前端按“农事消息”语义展示。 */
    private LocalDateTime publishTime;

    /** 发布人名称别名，优先取昵称，兜底取账号。 */
    private String publisherName;

    /** 更新时间，记录最后一次编辑或状态流转时间。 */
    private LocalDateTime updateTime;

    /** 是否已经超过截至时间，由查询层动态计算，不作为互斥业务状态持久化。 */
    private Boolean overdue;
}
